code/modules/mob/living/simple_animal/bot/turtle/turtle_cartridge.dm

```diff
@@ -0,0 +1,280 @@
+
+/obj/item/disk/turtle_cartridge/proc/reset()
+	line = 1
+	variables = list()
+
+/obj/item/disk/turtle_cartridge/proc/throw_error(err, part)
+	turtle?.speak("[err]:[part ? " " : ""][part ? part : ""] ON LINE [line]")
+	turtle?.beep("error")
+	stop()
+
+/obj/item/disk/turtle_cartridge/proc/is_varname(varname)
+	return varname_regex.Find(varname) && length(varname) <= 16
+
+/obj/item/disk/turtle_cartridge/proc/format_string(S)
+	for(var/varname in variables)
+		S = replacetext(S, "\[[varname]\]", "[variables[varname]]")
+	return S
+
+/obj/item/disk/turtle_cartridge/proc/process_val(val)
+	if(!isnull(text2num(val)))
+		return text2num(val)
+	if(val in variables)
+		return variables[val]
+	if(text2dir(val)) //direction vars are builtin, but can be overidden, see two lines above
+		return text2dir(val)
+
+	if(is_varname(val))
+		throw_error("UNDEFINED VAR", val)
+	else
+		throw_error("SYNTAX ERROR", val)
+
+/obj/item/disk/turtle_cartridge/proc/process_expression(list/args)
+	if(args.len == 1)
+		return process_val(args[1])
+	else if(args.len == 2)
+		switch(args[1])
+			if("NOT")
+				return !process_val(args[2])
+			if("ABS")
+				return abs(process_val(args[2]))
+			if("FLOOR")
+				return FLOOR(process_val(args[2]), 1)
+			if("CEIL")
+				return CEILING(process_val(args[2]), 1)
+			if("ROUND")
+				return round(process_val(args[2]))
+			if("NEG")
+				return -process_val(args[2])
+			if("SQRT")
+				return sqrt(process_val(args[2]))
+			else
+				throw_error("SYNTAX ERROR")
+	else if(args.len == 3)
+		switch(args[2])
+			if("ADD")
+				return process_val(args[1]) + process_val(args[3])
+			if("SUB")
+				return process_val(args[1]) - process_val(args[3])
+			if("MUL")
+				return process_val(args[1]) * process_val(args[3])
+			if("DIV")
+				if (process_val(args[3]) == 0) return throw_error("DIV BY 0")
+				return process_val(args[1]) / process_val(args[3])
+			if("MOD")
+				if (process_val(args[3]) == 0) return throw_error("DIV BY 0")
+				return process_val(args[1]) % process_val(args[3])
+			if("POW")
+				return process_val(args[1]) ** process_val(args[3])
+			if("MIN")
+				return min(process_val(args[1]), process_val(args[3]))
+			if("MAX")
+				return max(process_val(args[1]), process_val(args[3]))
+			if("LT")
+				return process_val(args[1]) < process_val(args[3])
+			if("LE")
+				return process_val(args[1]) <= process_val(args[3])
+			if("GT")
+				return process_val(args[1]) > process_val(args[3])
+			if("GE")
+				return process_val(args[1]) >= process_val(args[3])
+			if("EQ")
+				return process_val(args[1]) == process_val(args[3])
+			if("NE")
+				return process_val(args[1]) != process_val(args[3])
+			if("OR")
+				return process_val(args[1]) || process_val(args[3])
+			if("AND")
+				return process_val(args[1]) && process_val(args[3])
+			else
+				throw_error("SYNTAX ERROR")
+
+/obj/item/disk/turtle_cartridge/proc/process_line()
+	if(!lines[line])
+		stop()
+		return
+
+	turtle?.speak("PROCESSING LINE [lines[line]]")
```

Comment:
This should really be disabled by default unless the "coder" includes some sort of flag at the start of their code. Outside of debugging, I can't tell how this could be useful for anything other than spamming.

Reply:
Currently just my debug tool, will not be in prod